File: leetcode_python/Hash_table/reordered-power-of-2.py
"""

869. Reordered Power of 2
Medium

You are given an integer n. We reorder the digits in any order (including the original order) such that the leading digit is not zero.

Return true if and only if we can do this so that the resulting number is a power of two.

Example 1:

Input: n = 1
Output: true

Example 2:

Input: n = 10
Output: false

Constraints:

1 <= n <= 10^9

"""

# V0 

# V1 
# https://blog.csdn.net/fuxuemingzhu/article/details/82468653
# IDEA : 2**k 
# << PYTHON OPERATOR 
# https://wiki.python.org/moin/BitwiseOperators
# x << y : Returns x with the bits shifted to the left by y places (and new bits on the right-hand-side are zeros). This is the same as multiplying x by 2**y.
# DEMO 
# In [17]: for i in range(11):
#     ...:     print ( 1 << i)
#     ...:     
# 1
# 2
# 4
# 8
# 16
# 32
# 64
# 128
# 256
# 512
# 1024
# time = O((logn)^2) = O(1) due to n is a 32-bit number
# space = O(logn) = O(1)
class Solution(object):
    def reorderedPowerOf2(self, N):
        """
        :type N: int
        :rtype: bool
        """
        c = collections.Counter(str(N))
        return any(c == collections.Counter(str(1 << i)) for i in range(32))

# V2
# time = O((logn)^2) = O(1) due to n is a 32-bit number
# space = O(logn) = O(1)
import collections
class Solution(object):
    def reorderedPowerOf2(self, N):
        """
        :type N: int
        :rtype: bool
        """
        count = collections.Counter(str(N))
        return any(count == collections.Counter(str(1 << i)) for i in range(31))

File: leetcode_java/src/main/java/LeetCodeJava/HashTable/ReorderedPowerOf2.java
package LeetCodeJava.HashTable;

// https://leetcode.com/problems/reordered-power-of-2/

import java.util.Arrays;

/**
 *  869. Reordered Power of 2
 *  Medium
 *
 *  You are given an integer n. We reorder the digits in any order (including the
 *  original order) such that the leading digit is not zero.
 *
 *  Return true if and only if we can do this so that the resulting number is a
 *  power of two.
 *
 *  Example 1:
 *  Input: n = 1
 *  Output: true
 *
 *  Example 2:
 *  Input: n = 10
 *  Output: false
 *
 *  Constraints:
 *  1 <= n <= 10^9
 */
public class ReorderedPowerOf2 {

    // V0
    // IDEA: DIGIT COUNT SIGNATURE - two numbers are permutations of each other
    //       iff their sorted digit strings are equal. There are only 31 powers of
    //       2 within the int range, so compare n against all of them.
    /**
     * time = O((log n)^2) -> O(1), n is a 32-bit int
     * space = O(log n) -> O(1)
     */
    public boolean reorderedPowerOf2(int n) {

        String target = signature(n);

        for (int i = 0; i < 31; i++) {
            if (target.equals(signature(1 << i))) {
                return true;
            }
        }

        return false;
    }

    /** sorted digit string, e.g. 128 -> "128", 821 -> "128" */
    private String signature(int x) {
        char[] c = String.valueOf(x).toCharArray();
        Arrays.sort(c);
        return new String(c);
    }
}
